src/api/uploads/MultiputUpload.js

```diff
@@ -0,0 +1,367 @@
+        }
+
+        this.updateFirstUnuploadedPartIndex();
+
+        while (this.canStartMorePartUploads()) {
+            this.uploadNextPart();
+        }
+    };
+
+    /**
+     * Send a request to commit the upload.
+     * TODO: implement this
+     * 
+     * @private
+     * @return {void}
+     */
+    commitSession = (): void => {};
+
+    /**
+     * Commit response handler.  Succeeds the upload, retries the commit on 202/429/failure, or fails
+     * the upload.
+     * TODO: implement this
+     * 
+     * @private
+     * @return {void}
+     */
+    commitSessionSuccessHandler = (): void => {};
+
+    /**
+     * Commit error handler.
+     * Retries the commit or fails the multiput session.
+     * TODO: implement this
+     * 
+     * @private
+     * @return {void}
+     */
+    commitSessionErrorHandler = (): void => {};
+
+    /**
+     * Find first part in parts array that we can upload, and upload it.
+     * 
+     * @private
+     * @return {void}
+     */
+    uploadNextPart = (): void => {
+        for (let i = this.firstUnuploadedPartIndex; i < this.parts.length; i += 1) {
+            const part = this.parts[i];
+
+            if (part.state === PART_STATE_DIGEST_READY) {
+                // Update the counters here instead of uploadPart because uploadPart
+                // can get called on retries
+                this.partsDigestReady -= 1;
+                this.partsUploading += 1;
+                this.uploadPart(part);
+                break;
+            }
+        }
+    };
+
+    /**
+     * Upload given part
+     * TODO: implement this
+     * 
+     * @private
+     * @param {MultiputPart} part
+     * @return {void}
+     */
+    // eslint-disable-next-line
+    uploadPart = (part: MultiputPart): void => {};
+
+    /**
+     * Checks if upload pipeline is full
+     * TODO: support progressive hashing
+     * 
+     * @private
+     * @return {boolean}
+     */
+    canStartMorePartUploads = (): boolean => !this.ended && this.partsUploading < this.config.parallelism;
+
+    /**
+     * Functions that walk the parts array get called a lot, so we cache which part we should
+     * start work at to avoid always iterating through entire parts list.
+     * 
+     * @private
+     * @return {void}
+     */
+    updateFirstUnuploadedPartIndex = (): void => {
+        for (let i = this.firstUnuploadedPartIndex; i < this.parts.length; i += 1) {
+            const part = this.parts[i];
+
+            if (i === this.firstUnuploadedPartIndex && part.state === PART_STATE_UPLOADED) {
+                this.firstUnuploadedPartIndex = i + 1;
+            } else {
+                break;
+            }
+        }
+    };
+
+    /**
+     * After session is created and we know the part size, populate the parts
+     * array.
+     * 
+     * @private
+     * @return {void}
+     */
+    populateParts = (): void => {
+        this.partsNotStarted = Math.ceil(this.file.size / this.partSize);
+        this.parts = [];
```

Comment:
Remove this line, it's already initialized to an empty array in the constructor

Reply:
done